### PR TITLE
MAINT/ENH: Un-cythonize and create extension point for clocks

### PR DESCRIPTION
**Pull request by jnazaren**
Moves the `MinuteSimulationClock` from cython to python, and simplifies some of its logic. Also makes the class implement a new `Clock` interface (created with @ssanderson's `interface` module), which will allow us to create extension points for clocks in the near future. @llllllllll hinted that there may be a point at which we want to take into account the scheduling of a user's functions to make the yielding of values even more efficient, but that we're not quite ready to do that yet.

### CODE REVIEW

File: zipline/gens/clock.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+class SessionEvt(IntEnum):
+    BAR = 0
+    SESSION_START = 1
+    SESSION_END = 2
+    MINUTE_END = 3
+    BEFORE_TRADING_START_BAR = 4
+
+
+BAR = SessionEvt.BAR
```
[the remark this answers is not in the document]
Comment (jnazaren) [reply zipline/gens/clock.py]: That was just making it so that I wouldn't have to change a lot about the current imports used in other places

File: conda/python-interface/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  number: 0
+  script: python setup.py install --single-version-externally-managed --record=record.txt
+
+requirements:
+  build:
+    - python
+    - setuptools
+    - six
+    - typing >=3.5.2
+    - funcsigs >=1.0.2    [py<30]
```
Comment (richafrank) [inline conda/python-interface/meta.yaml]: Why are these required for the build (as opposed to run)?

File: zipline/extensions.py [hunk 3]
```diff
@@ -196,22 +196,22 @@ def load(interface, name):
 
 
 @curry
-def register(interface, name, custom_class):
+def register(interface, name, factory):
     """
-    Registers a class for retrieval by the load method
+    Registers a class factory for retrieval by the load method
```
[the remark this answers is not in the document]
Comment (jnazaren) [reply zipline/extensions.py]: With the changes I'm currently implementing (namely adding the ability for the user to specify `*args` and `**kwargs` when calling the load method, which will return a partial function), this will be a factory for instances or partial instances, meaning that a user can either register a method that fully sets up and returns an instance of the class, or register the class's constructor, which can then take extra arguments.

File: zipline/gens/clock.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+                yield counter, BAR
+                if self.minute_emission:
+                    yield counter, MINUTE_END
+
+                counter += pd.Timedelta(_nanos_in_minute)
+
+            yield market_close, SESSION_END
+
+
+@register(Clock, 'default')
```
[the remark this answers is not in the document]
Comment (jnazaren) [reply zipline/gens/clock.py]: See my response to your first question above. I'm making a few more changes to this, as well as the `load` function, which will allow for much more flexibility in terms of what a user can easily register. At this point, I'm only expecting for a `Clock` to be consumed by the `AlgorithmSimulator` (at least in Zipline), but I'm expecting this to be defined by an argument similar to the one for `blotter` in __main__. `blotter_class` is used in the creation of `TradingAlgorithm` to store the loaded `Blotter` instance (or partial instance) in the case where a user doesn't access `TradingAlgorithm` via the setup in run_algo.

File: conda/python-interface/meta.yaml (hunk 2 above)
Comment (jnazaren) [reply conda/python-interface/meta.yaml]: I'm not too sure about this one. Conda automatically added this into the recipe file for both build and run, so I trusted its judgement.

File: zipline/gens/clock.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jnazaren) [reply zipline/gens/clock.py]: At the moment, I don't think so, but I believe this was intended to be used in situations where a user's code creates a `TradingAlgorithm` on its own.

[the remark this answers is not in the document]
Comment (richafrank) [reply zipline/gens/clock.py]: Oh, I think we're generally trying to make `TradingAlgorithm` not a public interface anymore.

[the remark this answers is not in the document]
Comment (jnazaren) [reply zipline/gens/clock.py]: Ah ok, in that case it does seem unnecessary. I'll remove it.
